### PR TITLE
WIP deactivated user got 500 trying to login

### LINKED ISSUE #1631: deactivated user got 500 trying to login
user said:

> seems i found out a way to get "500 Internal Error Something went terribly wrong".. when https://www.agileventures.org/users/sign_in with my own credentials oliv...@gmail.com ... Terribly sorry, seems i disables my account and since then .. 500 internal error..

I found their user account using `u = User.with_deleted.find_by email: <email>` and then set the deleted_at column to nil and saved and hopefully that's fixed it, but ideally the system would tell the user they are deactivated and what to do rather than the 500 error ....

### PR DESCRIPTION
**Pull request by praveendhawan**
fixes #1631 

added a feature spec for duplicating the issue but could not get the test to fail

### CODE REVIEW

**Comment by tansaku** [comment]
I can confirm that that this does not fail as expected - I used byebug to dump the code and the user ...

```
→ bundle exec cucumber features/users/sign_in.feature:41
[Coveralls] Using SimpleCov's 'rails' settings.
Using the default profile...
@vcr
Feature: Sign in
  As an existing User
  So that I can use the systems all functions
  I want to be able to login

  Scenario: User is deactivated           # features/users/sign_in.feature:41
    Given I exist as a deactivated user   # features/step_definitions/user_steps.rb:72
      example@example.com
      changemesomeday
    And I am not logged in                # features/step_definitions/user_steps.rb:59
    When I sign in with valid credentials # features/step_definitions/user_steps.rb:95
      example@example.com
      changemesomeday

[201, 210] in /Users/tansaku/Documents/GitHub/AgileVentures/WebsiteOne/features/step_definitions/user_steps.rb
   201:   expect(page).to have_content "Signed out successfully."
   202: end
   203: 
   204: Then /^I see an invalid login message$/ do
   205:   byebug
=> 206:   expect(page).to have_content "Invalid email or password."
   207: end
   208: 
   209: Then /^I should (not |)see my name$/ do |should|
   210:   create_user
(byebug) page.body
"<!DOCTYPE html>\n<html>\n\n<head>\n  <script>\n//<![CDATA[\n\n      window._rails_env = \"test\"\n\n//]]>\n</script>  <meta charset=\"utf-8\"/>\n<meta http-equiv=\"X-UA-Compatible\" content=\"IE=edge\">\n<meta name=\"viewport\" content=\"width=device-width, initial-scale=1\">\n<meta name=\"keywords\" content=\"AgileVentures, pair programming, crowdsourced learning\">\n<meta name=\"description\" content=\"AgileVentures is a project incubator that stimulates and supports development of social innovations, open source and free software. We are also a community for learning and personal development with members from across the world with various levels of competence and experience in software development. We hold scrum meetings and pair programming sessions every day with participants from all time zones and on all levels. Come and join us.\">\n<meta name=\"author\" content=\"AgileVentures volunteers and contributors\">\n<meta name=\"copyright\" content=\"AgileVentures Nonprofit Ltd.\">\n<meta name=\"coverage\" content=\"Worldwide\">\n<meta name=\"distribution\" content=\"Global\">\n<meta name=\"application-name\" content=\"AgileVentures Official Website\">\n\n<meta property=\"og:title\" content=\"AgileVentures\"/>\n<meta property=\"og:type\" content=\"website\"/>\n<meta property=\"og:url\" content=\"http://www.example.com/users/sign_in\"/>\n<meta property=\"og:site_name\" content=\"AgileVentures\"/>\n<meta property=\"og:description\" content=\"AgileVentures is a project incubator that stimulates and supports development of social innovations, open source and free software. We are also a community for learning and personal development with members from across the world with various levels of competence and experience in software development. We hold scrum meetings and pair programming sessions every day with participants from all time zones and on all levels. Come and join us.\">\n<meta property=\"og:image\" content=\"http://www.example.com/assets/screenshot.png\">\n\n<meta name=\"twitter:card\" content=\"summary\">\n<meta name=\"twitter:title\" content=\"AgileVentures\"/>\n<meta name=\"twitter:url\" content=\"http://www.example.com/users/sign_in\">\n<meta name=\"twitter:site\" content=\"@AgileVentures\">\n<meta name=\"twitter:image:src\" content=\"http://www.example.com/assets/screenshot.png\">\n<meta name=\"twitter:description\" content=\"AgileVentures is a project incubator that stimulates and supports development of social innovations, open source and free software. We are also a community for learning and personal development with members from across the world with various levels of competence and experience in software development. We hold scrum meetings and pair programming sessions every day with participants from all time zones and on all levels. Come and join us.\">\n\n  <title>AgileVentures - WebsiteOne</title>\n  <link rel=\"stylesheet\" media=\"all\" href=\"/assets/application.css\" data-turbolinks-track=\"true\" />\n  <link rel=\"stylesheet\" media=\"screen\" href=\"https://fonts.googleapis.com/css?family=Lato:400,100,900\" />\n  <link rel=\"stylesheet\" media=\"screen\" href=\"https://fonts.googleapis.com/css?family=Ubuntu\" />\n  <script src=\"/assets/application.js\" data-turbolinks-track=\"true\"></script>\n  \n  <link rel=\"shortcut icon\" href=\"/favicon.ico?v=2\" />\n</head>\n\n\n\n<body lang=\"en-GB\">\n\n<div id=\"wrap\">\n  <header id=\"main_header\" class=\"masthead\">\n  <div class=\"container\">\n    <a class=\"navbar-brand\" href=\"/\"><img src=\"/assets/av-logo-inverse.png\"></a>\n\n    <div class=\"navbar-header navbar-inverse pull-right\">\n      <button type=\"button\" class=\"navbar-toggle\" data-toggle=\"collapse\" data-target=\".navbar-collapse\">\n        <span class=\"icon-bar\"></span>\n        <span class=\"icon-bar\"></span>\n        <span class=\"icon-bar\"></span>\n      </button>\n    </div>\n\n    <div id=\"account-controls\" class=\"hidden-xs pull-right\">\n        <div id=\"login-signup\">\n          <a href=\"/users/sign_in\">Log in</a>\n          <a id=\"signup\" href=\"/users/sign_up\">Sign up</a>\n        </div>\n    </div>\n  </div>\n</header>\n\n<div class=\"navbar navbar-inverse navbar-static-top collapse navbar-collapse hidden-sm\" id=\"nav\">\n  <div class=\"container\">\n    <div class=\"row\">\n      <div class=\"hidden-sm col-xs-3 col-md-2 navbar-brand-div\">\n        <a id=\"affix-nav-brand\" class=\"navbar-brand\" href=\"/\">\n          <img src=\"/assets/av-logo-inverse.png\">\n        </a>\n      </div>\n      <div class=\"col-xs-12 col-sm-9 col-md-10\">\n        <div class=\"\">\n          <ul class=\"nav navbar-nav nav-justified\">\n            <li><a class=\"\" href=\"/about-us\">About</a></li>\n            <li><a href=\"/projects\">Projects</a></li>\n            <li><a href=\"/users\">Members</a></li>\n            <li><a href=\"/pricing\">Premium</a></li>\n            <li class=\"dropdown\">\n              <a href=\"#\" id=\"events-dropdown\" data-toggle=\"dropdown\">Events <span class=\"caret\"></span></a>\n              <ul class=\"dropdown-menu custom_menu\">\n                <a class=\"custom_menu_link\" href=\"/events\">\n                  <li><i class=\"fa fa-calendar-o custom_menu_item\"></i>Upcoming events</li>\n                </a>\n                <a class=\"custom_menu_link\" href=\"/scrums\">\n                  <li><i class=\"fa fa-play custom_menu_item\"></i>Past scrums</li>\n                </a>\n                <a class=\"custom_menu_link\" href=\"/hangouts\">\n                  <li><i class=\"fa fa-bookmark-o custom_menu_item\"></i>Past events</li>\n                </a>\n                <a class=\"custom_menu_link\" href=\"/events/new\">\n                  <li><i class=\"fa fa-users custom_menu_item\"></i>Create event</li>\n                </a>\n              </ul>\n            </li>\n            <li><a class=\"\" href=\"/getting-started\">Getting started</a></li>\n\n            <li class=\"dropdown\" id=\"google-search-dropdown\">\n              <a href=\"#\" id=\"google_search\" data-toggle=\"dropdown\"><i class=\"fa fa-search\"></i></a>\n              <ul class=\"dropdown-menu\" id=\"google-search-wrapper\">\n                <gcse:search></gcse:search>\n              </ul>\n            </li>\n\n                <li class=\"visible-xs\">\n                  <a href=\"/users/sign_in\">Log in</a>\n                </li>\n                <li class=\"visible-xs\">\n                  <a id=\"signup\" href=\"/users/sign_up\">Sign up</a>\n                </li>\n          </ul>\n        </div>\n      </div>\n    </div>\n  </div>\n    <div id=\"flash-container\">\n    <div id=\"flash-display\">\n        <div id=\"flash_alert\"\n             class=\"alert alert-danger\">\n          <a class=\"close\" data-dismiss=\"alert\">&#215;</a>\n          <p style=\"margin: 0 auto;\">\n            Invalid email or password.\n          </p>\n        </div>\n    </div>\n  </div>\n\n</div>\n\n  <div id=\"main\">\n      <div class=\"container\">\n    <div class=\"col-lg-9\">\n      <div class=\"container\">\n<div class=\"col-md-5 col-md-offset-2\">\n  <h1 class=\"text-center\"> Log In  </h1>\n  <div class=\"row\">\n        <div class=\"col-xs-6 col-sm-6 col-md-6\">\n            <div data-no-turbolink>\n      <a class=\"btn btn-block btn-social btn-github btn-lg\"  href=\"/auth/github?origin=%2F\">\n        <i class=\"fa fa-github-alt\"></i> with GitHub\n      </a>\n    </div>\n\n        </div>\n        <div class=\"col-xs-6 col-sm-6 col-md-6\">\n            <div data-no-turbolink>\n      <a class=\"btn btn-block btn-social btn-gplus btn-lg\"  href=\"/auth/gplus?origin=%2F\">\n        <i class=\"fa fa-google\"></i> with Google\n      </a>\n    </div>\n\n        </div>\n  </div>\n  <h4 class=\"text-center\">Don't have an account? <a href=\"/users/sign_up\">Sign up</a></h4>\n  <h4 class=\"text-center\"><a href=\"/users/password/new\">Forgot your password?</a></h4>\n\n  <form class=\"new_user\" id=\"new_user\" action=\"/users/sign_in\" accept-charset=\"UTF-8\" method=\"post\"><input name=\"utf8\" type=\"hidden\" value=\"&#x2713;\" />\n      <fieldset>\n        <div class='form-group'>\n          <input autofocus=\"autofocus\" placeholder=\"Enter email\" class=\"form-control input-lg\" type=\"email\" value=\"example@example.com\" name=\"user[email]\" id=\"user_email\" />\n        </div>\n        <div class='form-group'>\n          <input placeholder=\"Enter password\" class=\"form-control input-lg\" type=\"password\" name=\"user[password]\" id=\"user_password\" />\n        </div>\n            <div class='form-group' style=\"display:inline-flex;\">\n              <input name=\"user[remember_me]\" type=\"hidden\" value=\"0\" /><input type=\"checkbox\" value=\"1\" name=\"user[remember_me]\" id=\"user_remember_me\" /> <label class=\"custom-label\" style=\"margin-left: 5px;\" for=\"user_remember_me\">Remember me</label>\n            </div>\n        <div class='form-group'>\n          <input type=\"submit\" name=\"commit\" value=\"Sign in\" class=\"btn btn-primary btn-lg btn-block\" />\n        </div>\n        \n      </fieldset>\n</form></div>\n</div>\n\n    </div>\n    <div class=\"col-lg-3\">\n      <div id=\"sponsorsBar\" class=\"hidden-sm hidden-xs hidden-md\" >\n  <div class=\"gold sponsorsTier text-center\">\n    <h3>Supporters</h3>\n    <ul>\n      <li>\n        <a href=\"https://drie.co/\" class=\"sponsorMedal\">\n          <div class=\"sponsorStandinMedal\">\n            <img alt=\"drie\" src=\"/assets/sponsors/drie.png\">\n          </div>\n        </a>\n      </li>\n      <li>\n        <a href=\"http://craftacademy.se/english\" target='new' class=\"sponsorMedal\">\n          <div class=\"sponsorStandinMedal\">\n            <img alt=\"Craft Academy\" src=\"/assets/sponsors/craft_banner_2.png\">\n          </div>\n        </a>\n      </li>\n      <li>\n        <a href=\"http://www.mentive.co/topics/?utm_campaign=coursepage&utm_medium=banner&utm_source=agileventures\" target='new' class=\"sponsorMedal\">\n          <div class=\"sponsorStandinMedal\">\n            <img alt=\"Mentive\" src=\"/assets/sponsors/mentive.jpg\">\n          </div>\n        </a>\n      </li>\n      <li>\n        <a href=\"http://www.jetbrains.com/ruby/specials/rubymine/ruby_and_rails_ide.jsp\" class=\"sponsorMedal\">\n          <div class=\"sponsorStandinMedal\">\n            <img alt=\"RubyMine 8\" src=\"/assets/sponsors/RubyMine_sidebar.png\">\n          </div>\n        </a>\n      </li>\n      <li>\n        <span class=\"sponsorStandinMedal\" id='ourSponsors'>\n          <a class=\"btn btn-be-a-supporter\" href=\"/sponsors\">Become a supporter</a>\n        </span>\n      </li>\n    </ul>\n  </div>\n</div>\n\n    </div>\n  </div>\n  </div>\n</div>\n\n<footer id=\"footer\">\n  <div class=\"container\">\n  <div class=\"col-sm-7\">\n    <div id=\"upper-section\" class=\"row\">\n      <div class=\"footer-info col-xs-6 col-sm-4\">\n        <ul>\n          <h3>Learn More</h3>\n          <li><a href=\"/about-us\">About Us</a></li>\n          <li><a href=\"/getting-started\">Getting Started</a></li>\n          <li><a href=\"/dashboard\">Dashboard</a></li>\n          <li><a href=\"http://nonprofits.agileventures.org/blog/\">Blog</a> </li>\n          <li><a href=\"http://www.agileventures.org/press-kit\">Press Kit</a></li>\n        </ul>\n      </div>\n\n      <div class=\"footer-info col-xs-6 col-sm-4\">\n        <ul>\n          <h3>Social</h3>\n          <li>\n            <a class=\"facebook\" href=\"https://www.facebook.com/agileventures\" target=\"_blank\" style=\"white-space: nowrap\">\n              <i class=\"fa fa-facebook fa-custom-social\"></i>&nbsp;Facebook\n            </a>\n          </li>\n          <li>\n            <a class=\"twitter\" href=\"https://twitter.com/AgileVentures\" target=\"_blank\" style=\"white-space: nowrap\">\n              <i class=\"fa fa-twitter fa-custom-social\"></i>&nbsp;Twitter\n            </a>\n          </li>\n        </ul>\n      </div>\n\n      <div class=\"footer-info col-xs-12 col-sm-4\">\n        <ul>\n          <h3>Our Sponsors</h3>\n          <li>\n            <a href=\"https://drie.co/\" target=\"_blank\">drie</a>\n          </li>\n          <li>\n            <a href=\"http://craftacademy.se/english\" target=\"_blank\">Craft Academy</a>\n          </li>\n          <li>\n            <a href=\"http://www.mentive.co/\" target=\"_blank\">Mentive</a>\n          </li>\n          <li>\n            <a href=\"https://www.jetbrains.com/ruby/\" target=\"_blank\">RubyMine</a>\n          </li>\n          <li>\n            <a style=\"font-size: 0.9em;\" href=\"/sponsors\">Becoming a sponsor</a>\n          </li>\n        </ul>\n      </div>\n    </div>\n\n    <div class=\"row\">\n      <h3><img src=\"/assets/av-logo-inverse.png\"/></h3>\n      <p>AgileVentures - Crowdsourced Learning</p>\n    </div>\n  </div>\n\n  <div class=\"col-sm-5 contact\">\n    <h3 class=\"footer_header\">\n      Contact us\n    </h3>\n    <p>Send a traditional email to <a href=\"mailto:info@agileventures.org\">info@agileventures.org</a>.</p>\n    <p>We are a Charitable Incorporated Organisation (CIO) in the UK. Ref #1170963</p>\n  </div>\n</div>\n\n</footer>\n\n</body>\n</html>\n"
(byebug) User.find_by email: 'example@example.com'
nil
(byebug) User.with_deleted.find_by email: 'example@example.com'
#<User id: 2, email: "example@example.com", encrypted_password: "$2a$04$nFvmgmrSlZeS1nW4lXgzfuZBl/0wjLmyB9V6wSXHglA...", reset_password_token: nil, reset_password_sent_at: nil, remember_created_at: nil, sign_in_count: 0, current_sign_in_at: nil, last_sign_in_at: nil, current_sign_in_ip: nil, last_sign_in_ip: nil, created_at: "2017-09-04 15:58:06", updated_at: "2017-09-04 15:58:07", first_name: "Anders", last_name: "Persson", display_email: nil, youtube_id: nil, slug: "slug-ma", display_profile: true, latitude: nil, longitude: nil, country_name: "Sweden", city: nil, region: nil, youtube_user_name: nil, github_profile_url: nil, display_hire_me: nil, bio: "Magnam explicabo exercitationem recusandae et nost...", receive_mailings: true, country_code: nil, timezone_offset: nil, status_count: 0, deleted_at: "2017-09-04 00:00:00">
(byebug) 

```

I think the options here are that either this has already been fixed, or it is an issue with a difference between test and production environments.

The next thing I would probably do is try to replicate manually on the develop server at https://develop.agileventures.org, but you'd need console access to delete a user - although actually not - you could create a user then and then delete through the interface ...

**Comment by praveendhawan** [comment]
@tansaku As suggested I went to [https://develop.agileventures.org](https://develop.agileventures.org) but could not get past the sign-up page as all the registration method were giving me some kind of errors. Google and Github were showing  `Redirect URI mismatch error` and when I tried to sign-up via my email, it showed me a notice saying `verification failed`.

**Comment by tansaku** [comment]
@praveendhawan ah yes, sorry - want to give https://staging.agilenentures.org a try - I must get those develop server login issues sorted ...

**Comment by praveendhawan** [comment]
@tansaku The staging URL is also not working. I am getting "This site can't be reached" error.
screen shot 2017-09-05 at 16 51 17

**Comment by tansaku** [comment]
sorry - mistype - try https://staging.agileventures.org/

**Comment by praveendhawan** [comment]
Now I am seeing 500 after deactivating my account. Steps to reproduce
1) Login via GitHub/Google
2) Deactivate your account
3) Try to login back using the same GitHub/Google profile.

screen shot 2017-09-05 at 17 35 51

**Comment by tansaku** [comment]
yes, well done @praveendhawan - so it looks like this is a bug that we are only seeing in production mode ...

you could try and replicate it on your local machine by running your local instance in production mode, but I warn you that it's a little painful to get it sorted - it will start asking for airbrake env vars and other things

the other thing to is to look through the settings differences between dev, production and test in the all - there's a big load of settings files in the config dir, and you might poke around for something that might affect this.

it would also make sense to see if you can replicate running locally in development mode ...

**Comment by praveendhawan** [comment]
@tansaku I think the problem was not in the production environment as I was able to reproduce the problem in test environment also. I wrote a test and it exited with following error message 
`ERROR:  duplicate key value violates unique constraint "index_users_on_email"
      DETAIL:  Key (email)=(mock@email.com) already exists.
       (PG::UniqueViolation)
      ./app/controllers/authentications_controller.rb:100:in `create_new_user_with_authentication'
      ./app/controllers/authentications_controller.rb:17:in `create'`

**Comment by tansaku** [comment]
ah good work @praveendhawan - so we are only seeing this error when the user is first created via google? or also via github?

do we not get the error on staging with a user created via email?

**Comment by praveendhawan** [comment]
@tansaku  Yes we are only getting this 500 error when the user is signed up via **Google/Github** and try to sign in after deactivating his account via - **Google/Github or Email/Password**. But if he signed up via email
he will see `Invalid credentials` flash message.

**Comment by praveendhawan** [comment]
@tansaku  Just wanted a clarification regarding the above change you requested
I don't want to put the code in `AppicationController` as it is only specific to `AuthenticationsControler` and `RegistrationsController`. Also, I can see there is something called `UsersHelper`. By module mixin, you meant to put the code in that helper class or should I make a new concern in controllers folder with these methods?

**Comment by tansaku** [comment]
@praveendhawan sure - ApplicationsController is not ideal, but it would be fine for a quick fix.  UsersHelper is for view components, you could pull things into a concern, which is maybe the cleanest, but up to you.

Feel free to make a refactoring ticket to explore.  For me the quick way to get this released is to pull the duplicated code into the ApplicationsController, but if you have time to quickly make a concern please go ahead

File: app/controllers/authentications_controller.rb [hunk 1]
```diff
@@ -1,4 +1,6 @@
 class AuthenticationsController < ApplicationController
+  include UsersHelper
```
Comment (tansaku) [inline app/controllers/authentications_controller.rb]: do we need to add the UsersHelper here?

Comment (praveendhawan) [reply app/controllers/authentications_controller.rb]: @tansaku I saw your comment after doing changes. I will revert it back to move the method to a concern.

**Comment by tansaku** [comment]
@praveendhawan - looks good - great work

btw, please don't feel the need to add additional controller specs when we have test coverage from acceptance tests - I personally can't see any value from controller specs.

Also, not sure why you were adding Userhelper specs in this PR, but this is an important fix and we should get it out, so I'm going to merge - many thanks again :-)

**Pull request merged**